lib/storage/gcs.js

```diff
@@ -0,0 +1,152 @@
+
+module.exports = function() {
+  var contentTypes;
+  var client;
+  var cachingTime;
+  var https;
+  var bucketName;
+  var endpoint;
+  var defaultTypes;
+  var noProtoEndpoint;
+
+  var self = {
+    init: function (options, callback) {
+      if (!(process.env.GOOGLE_APPLICATION_CREDENTIALS)) {
+        throw "GOOGLE_APPLICATION_CREDENTIALS not set in env, cannot proceed";
+      }
+      // Ultimately the result will look like https://storage.googleapis.com/[BUCKET_NAME]/[OBJECT_NAME]
+      // The rest is based mostly on s3 knox surmises.
+      endpoint = 'storage.googleapis.com';
+      if (options.endpoint) {
+        endpoint = options.endpoint;
+        if (!endpoint.match(/^https?:/)) {
+          var defaultSecure = ((!options.port) || (options.port === 443));
+          var secure = options.secure || defaultSecure;
+          var port = options.port || 443;
+          var protocol = secure ? 'https://' : 'http://';
+          if (secure && (port === 443)) {
+            port = '';
+          } else if ((!secure) && (port === 80)) {
+            port = '';
+          } else {
+            port = ':' + port;
+          }
+          endpoint = protocol + endpoint + port;
+        }
+      }
+      // The storage client auth relies on the presence of the service account
+      // file path expressed in the environment variable
+      // GOOGLE_APPLICATION_CREDENTIALS and, of course, the presence of such file.
+      //
+      //
+      // See https://cloud.google.com/docs/authentication/getting-started
+      client = new Storage();
+      bucketName = options.bucket;
+      defaultTypes = require(__dirname + "/contentTypes.js");
+      if (options.contentTypes) {
+        _.extend(contentTypes, defaultTypes, options.contentTypes);
+      } else {
+        contentTypes = defaultTypes;
+      }
+      defaultTypes = require(__dirname + "/contentTypes.js");
+      if (options.contentTypes) {
+        _.extend(contentTypes, defaultTypes, options.contentTypes);
+      } else {
+        contentTypes = defaultTypes;
+      }
+      https = options.https;
+      cachingTime = options.cachingTime;
+      return callback(null);
+    },
+
+    copyIn: function(localPath, path, options, callback) {
+      async.waterfall([
+        function(callbackOne) {
+          var cleanPath = cleanKey(path);
+          var ext = extname(cleanPath);
+          if (ext.length) {
+            ext = ext.substr(1);
+          }
+          var contentType = contentTypes[ext];
+          if (!contentType) {
+            console.log(`unknown extension ${ext}, using mime octet-stream`);
+            contentType = 'application/octet-stream';
+          }
+
+          let cacheControl = 'no-cache';
+          if (cachingTime) {
+            cacheControl = 'public, max-age=' + cachingTime;
+          }
+          callbackOne(null, {
+            destination: cleanPath,
+            gzip: true,
+            metadata: {
+              cacheControl: cacheControl,
```

Comment:
Do the capitalization differences between cacheControl and ContentType make sense? Honestly not sure.

Reply:
I took this directly from the Google documentation for the client. I will find a definitive answer and report back.